src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+
+                return new FilteredCompletionModel(
+                    ImmutableArray<CompletionItemWithHighlight>.Empty, selectedItemIndex: 0,
+                    _data.SelectedFilters, selection, centerSelection: true, uniqueItem: null);
+            }
+
+            private ImmutableArray<CompletionFilterWithState> GetUpdatedFilters(IReadOnlyList<MatchResult<VSCompletionItem>> items)
+            {
+                var showCompletionItemFilters = _globalOptions.GetOption(CompletionViewOptions.ShowCompletionItemFilters, _document?.Project.Language);
+                if (!showCompletionItemFilters)
+                    return ImmutableArray<CompletionFilterWithState>.Empty;
+
+                // See which filters might be enabled based on the typed code
+                using var _ = PooledHashSet<CompletionFilter>.GetInstance(out var textFilteredFilters);
+                textFilteredFilters.AddRange(items.SelectMany(n => n.EditorCompletionItem.Filters));
+
+                // When no items are available for a given filter, it becomes unavailable.
+                // Expanders always appear available as long as it's presented.
+                return _data.SelectedFilters.SelectAsArray(n => n.WithAvailability(n.Filter is CompletionExpander || textFilteredFilters.Contains(n.Filter)));
+            }
+
+            /// <summary>
+            /// Given multiple possible chosen completion items, pick the one that has the
+            /// best MRU index, or the one with highest MatchPriority if none in MRU.
+            /// </summary>
+            private static RoslynCompletionItem GetBestCompletionItemBasedOnMRU(
+                ImmutableArray<RoslynCompletionItem> chosenItems, ImmutableArray<string> recentItems)
+            {
+                Debug.Assert(chosenItems.Length > 0);
+
+                // Try to find the chosen item has been most recently used.
+                var bestItem = chosenItems[0];
+                for (int i = 1, n = chosenItems.Length; i < n; i++)
+                {
+                    var chosenItem = chosenItems[i];
+                    var mruIndex1 = GetRecentItemIndex(recentItems, bestItem);
+                    var mruIndex2 = GetRecentItemIndex(recentItems, chosenItem);
+
+                    if ((mruIndex2 < mruIndex1) ||
+                        (mruIndex2 == mruIndex1 && !bestItem.IsPreferredItem() && chosenItem.IsPreferredItem()))
+                    {
+                        bestItem = chosenItem;
+                    }
+                }
+
+                // If our best item appeared in the MRU list, use it
+                if (GetRecentItemIndex(recentItems, bestItem) <= 0)
+                {
+                    return bestItem;
+                }
+
+                // Otherwise use the chosen item that has the highest
+                // matchPriority.
+                for (int i = 1, n = chosenItems.Length; i < n; i++)
+                {
```

Comment:
```suggestion
                // Otherwise use the chosen item that has the highest matchPriority.
                for (int i = 1, n = chosenItems.Length; i < n; i++)
                {
```